File: code/python/剑指offer55-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        def balance(root):
            if root is None: return 0,True
            if root.left is not None:
                left,l = balance(root.left)
            else:
                left,l = 0,True
            if root.right is not None:
                right,r = balance(root.right)
            else:
                right,r = 0,True
            return max(left,right)+1,l and r and abs(left-right)<=1
        return balance(root)[1]

File: code/java/剑指offer55-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        return balance(root)!=-1;
    }
    public int balance(TreeNode root){
        if(root==null) return 0;
        int left = balance(root.left);
        if(left==-1) return -1;
        int right = balance(root.right);
        if(right==-1) return -1;
        return Math.abs(left-right)<=1?Math.max(left,right)+1:-1;
    }
}